### PR TITLE
Llama3.2-Vision: Add reference submodule and tests

### PR DESCRIPTION
**Pull request by cglagovichTT**
### Ticket
#13368

### Problem description
The multimodal reference code was not tracked in git. We now have a fork of it we should use as reference code in tests and demos https://github.com/tenstorrent/llama-models/tree/main

### What's changed
- Add llama-models as a submodule
- Move TT multimodal modules and tests into their own folders under `models/demos/llama3`
- Added tests to t3k unit, t3k frequent, and t3k demo. For each pipeline, I run the tests on a 1x2 mesh and a 1x8 mesh.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/11500543568
- [x] 
  - most recent t3k CI after hang fix was merged https://github.com/tenstorrent/tt-metal/actions/runs/11555134057
  - t3k CI after python dep changes https://github.com/tenstorrent/tt-metal/actions/runs/11558194724
  - Just One More CI run after removing `llama_models` as a submodule https://github.com/tenstorrent/tt-metal/actions/runs/11574438253
  - https://github.com/tenstorrent/tt-metal/actions/runs/11575070877

### CODE REVIEW

**Comment by cglagovichTT** [comment]
@tt-rkim could you check out `.gitmodules` to make sure that looks alright?

**Review by tt-rkim — APPROVED** [review]

**Comment by tt-rkim** [comment]
Looks good

Where are we running the `multimodal` tests exactly on CI?

**Comment by cglagovichTT** [comment]
> Where are we running the `multimodal` tests exactly on CI?
I'd like to add the multimodal tests to CI after I experiment locally to see if I can tighten up the PCC bounds. I'll make a separate PR to put the tests in CI.

**Comment by tt-rkim** [comment]
sounds good, so these reference files are not used anywhere in CI currently? Or are they used for other parts of 3.2?

**Comment by mtairum** [comment]
@cglagovichTT Shouldn't this PR also include the tests?

This is the PR that fully includes multimodal llama, so it's good practice to have it all good to go. 
But I'm ok with either approach.

**Review by mtairum — APPROVED** [review]

**Comment by cglagovichTT** [comment]
Due to popular demand I will add the tests to this PR :)

**Comment by tt-rkim** [comment]
THANK YOU SIR

File: models/demos/llama3/tt/model_config.py
Comment (cglagovichTT) [inline models/demos/llama3/tt/model_config.py:211]: @yieldthought is it going to be an issue in the non-vision attention modules to have fp32 acc here?

**Comment by cglagovichTT** [comment]
I'm encountering an issue where the python packages required in the new submodule `llama-models` aren't included in the tt-metal python env. The following are required packages:
```
blobfile
fairscale
jinja2
json-strong-typing
tiktoken
torch
pydantic
pydantic_core
```


there may be more that we're missing -- I'll check on a clean build. @tt-rkim

**Comment by mtairum** [comment]
@cglagovichTT You should be able to add them to the python env list:
`tt_metal/python_env/requirements-dev.txt`

**Comment by cglagovichTT** [comment]
Added them. Found that only `pydantic` and `pydantic_core` are actually needed to run the tests.

**Review by esmalTT — APPROVED** [review]
LGTM 🔥 🔥 🔥 (once CI is passing 🥲 )

**Comment by ttmchiou** [comment]
should we still be concerned with this segfault on llama?

**Comment by cglagovichTT** [comment]
> should we still be concerned with this segfault on llama?

It seems to be non-deterministic. I am able to repro it on main locally, so it's not an issue specific to this branch.

**Review by ttmchiou — APPROVED** [review]

File: tt_metal/python_env/requirements-dev.txt
Comment (TT-billteng) [inline tt_metal/python_env/requirements-dev.txt:60]: we really need to find ways to create separate dependencies suites for different use cases
If this is only needed for llama, there should be a separate pipenv the user creates

Comment (TT-billteng) [reply tt_metal/python_env/requirements-dev.txt:60]: Which file is pydantic used? I don't see a reference to it

Comment (cglagovichTT) [reply tt_metal/python_env/requirements-dev.txt:60]: it's required by `models/demos/llama3/reference/llama_models`

**Comment by cglagovichTT** [comment]
Think I caught a new issue in my CI run after merging main. https://github.com/tenstorrent/tt-metal/actions/runs/11505869140/job/32029208090

```
> /home/cglagovich/tt-metal/models/demos/llama3/demo/demo.py(351)run_llama3_demo()
-> trace_id = ttnn.begin_trace_capture(mesh_device, cq_id=0)
(Pdb)
> /home/cglagovich/tt-metal/models/demos/llama3/demo/demo.py(353)run_llama3_demo()
-> decode_input = ttnn.unsqueeze_to_4D(tt_embd(tt_out_tok))
(Pdb)
                 Always | FATAL    | Enqueue Read Buffer cannot be used with tracing
libc++abi: terminating due to uncaught exception of type std::runtime_error: TT_FATAL @ ../tt_metal/impl/dispatch/command_queue.cpp:1902: !this->manager.get_bypass_mode()
info:
Enqueue Read Buffer cannot be used with tracing
backtrace:
 --- /home/cglagovich/tt-metal/ttnn/ttnn/_ttnn.so(+0x3c9038) [0
```

Checking t3k demo on main: https://github.com/tenstorrent/tt-metal/actions/runs/11507082924

File: tt_metal/python_env/requirements-dev.txt
Comment (TT-billteng) [reply tt_metal/python_env/requirements-dev.txt:60]: why is it listed here then?

File: models/demos/llama3/reference/llama_models
Comment (TT-billteng) [inline models/demos/llama3/reference/llama_models:1]: Given that this submodule has a method to install the dependencies, can you dynamically install it whenever these tests are run? We should not be including everything under the sun in this repo

**Review by TT-billteng — CHANGES_REQUESTED** [review]
see my comments in regards how we fetch dependencies of the `llama_models` submodule

File: tt_metal/python_env/requirements-dev.txt
Comment (TT-billteng) [reply tt_metal/python_env/requirements-dev.txt:60]: you can install the pip wheel before you run the test in the GH workflow. That would remove the need for you deal with dependencies completely right?

**Comment by cglagovichTT** [comment]
The hang blocking this PR was resolved. https://github.com/tenstorrent/tt-metal/issues/14310#issuecomment-2439770070

A change to reshape uncovered a deadlock in async runtime. The bad commit was reverted, and a fix to runtime is in PR.

File: tests/scripts/t3000/run_t3000_demo_tests.sh
Comment (tt-rkim) [inline tests/scripts/t3000/run_t3000_demo_tests.sh:92]: It is possible that because we're using cached python dependencies, the llama submodule dependencies may actually make their way into the dependencies cache on GitHub.

So, if a T3K runner that:
- will run these llama tests
- were to be the next runner on main that runs changes to any of the paths for the cache key (ie. `tt_metal/python_env/requirements-dev.txt`, `pyproject.toml` etc)

then these llama submodule pip dependencies will make it into the cache.

Now, this should be an exceedingly rare event, as single-card pipelines are generally run before T3K ones when developers are working on CI. So, I'm okay with this, but I'm just pointing out that technically bloat / potential conflicts could make it into the cache.

The fix in that case is to just blow away the cache entries and re-run pipelines. So, not a complicated situation to get ourselves out of, anyway.

cc: @ttmchiou @TT-billteng

**Review by tt-rkim — APPROVED** [review]
LGTM, make sure to run all those pipelines before merging, of course

**Review by yieldthought — APPROVED** [review]
We should refactor the layernorm/distributed rmsnorm split at some point but it doesn't have to be now

**Review by TT-billteng — APPROVED** [review]
love it!

**Pull request merged**
